### PR TITLE
e2e framework

### PR DESCRIPTION
**Pull request by sjenning**
First step toward addressing #698, this PR introduces an e2e framework based on the kube implementation.

The `util.go` in kube has many more helper functions, but I figure we can bring those over on an as-needed basis as that file is huge in kube with lots of dependencies.

`make test-e2e` will run the e2e tests, which only contain a sample test case that always passes right now as an example.

I broke this PR into three commits
- import of the ginkgo and gomega deps
- add the e2e framework
- add an broker e2e test as an example

@derekwaynecarr @pmorie @MHBauer @arschles

### CODE REVIEW

**Comment by pmorie** [comment]
> The util.go in kube has many more helper functions, but I figure we can bring those over on an as-needed basis as that file is huge in kube with lots of dependencies.

Yeah, we could also make a new kube/e2e repo and put stuff there and consume it from both this repo and k8s

**Comment by MHBauer** [comment]
I think you need to checkin the changes to `glide.lock` as well.

**Comment by pmorie** [comment]
@sjenning how about an extremely basic E2E along with this that adds a Broker and waits for it to become ready?

**Comment by sjenning** [comment]
@MHBauer I've pushed a new commit that is the result of a `glide update -v`, since the lock file is only regenerated when a full dependency tree is calculated.

There are some other deps that are getting updated for reasons unknown to me.  Doesn't seem to break anything though.

**Comment by sjenning** [comment]
@pmorie so in order for the broker resource to become ready, there will need to be a broker pod running.

so you are asking for:
- create broker pod in kube
- wait for it to be running
- create broker service in kube
- create broker resource in sc with URL pointing to broker pod
- wait for it to become ready

That right?

**Comment by sjenning** [comment]
@pmorie broker e2e is now included, PTAL

**Comment by pmorie** [comment]
@sjenning yep, that's exactly what i was looking for, thanks.

**Comment by MHBauer** [comment]
How do I run this?

**Comment by MHBauer** [comment]
Is any of this copied from upstream? Can we use it directly vs doing the copy?

I guess we can use ginkgo in our unit tests after this. That might be nice.

**Comment by sjenning** [comment]
@MHBauer the upstream framework is much more complicated.  Plus we need to have both a kube client and service catalog client in our case and we don't want upstream e2e changes wrecking our e2e on each rebase IMHO.

If you want to run it
```
make bin/e2e.test
bin/e2e.test -kubernetes-config=path/to/kube.config -service-catalog-config=path/to/servicecatalog.config
```
Alternatively, you can set the config files with `KUBECONFIG` and `SERVICECATALOGCONFIG`, or if you have them as two contexts in the same config, use `-kubernetes-context` and `-service-catalog-context`

`make test-e2e` assumes that you have `KUBECONFIG` and `SERVICECATALOGCONFIG` set for now.

**Comment by MHBauer** [comment]
Put that comment into the `docs/devguide.md`, please. I'll give it a run through and report back.

**Comment by MHBauer** [comment]
Does kubernetes have to be running?

Does the service catalog have to be running?

**Comment by sjenning** [comment]
> Does kubernetes have to be running?
>
>Does the service catalog have to be running?

 Yes and yes

File: test/e2e/broker.go [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+		ObjectMeta: metav1.ObjectMeta{
+			Name: name,
+		},
+		Spec: v1alpha1.BrokerSpec{
+			URL: url,
+		},
+	}
+}
+
+func newTestBrokerPod(name string) *v1.Pod {
```
Comment (pmorie) [inline test/e2e/broker.go]: One thing I have often considered in upstream e2es is using the serialized form of a versioned API inside a multi-line string and deserializing in the test.  That way you could:

- easily create an e2e from a yaml snippet
- easily mutate an e2e to debug (it kind of sucks having to modify go code relative to tweaking a yaml)

Comment (pmorie) [inline test/e2e/broker.go]: I might, in a follow-up, prototype that and try to find the right programming model.

File: test/e2e/broker.go [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+	AfterEach(func() {
+		By("Deleting the user broker pod")
+		err := f.KubeClientSet.CoreV1().Pods(f.Namespace.Name).Delete(brokerName, nil)
+		Expect(err).NotTo(HaveOccurred())
+		By("Deleting the user broker service")
+		err = f.KubeClientSet.CoreV1().Services(f.Namespace.Name).Delete(brokerName, nil)
+		Expect(err).NotTo(HaveOccurred())
+	})
+
+	It("should become ready", func() {
```
Comment (pmorie) [inline test/e2e/broker.go]: I like the programming model here.

**Comment by jpeeler** [comment]
I've seen this error twice now, but perhaps I'm doing something wrong:
```
bin/e2e.test
I0428 11:09:24.153464    5331 e2e.go:44] Starting e2e run "aa0d1d83-2c24-11e7-9523-54ee7511a25f" on Ginkgo node 1
Running Suite: Service Catalog e2e suite
========================================
Random Seed: 1493392164 - Will randomize all specs
Will run 1 of 1 specs

[service-catalog] Broker 
  should become ready
  /home/jpeeler/src/go/src/github.com/kubernetes-incubator/service-catalog/test/e2e/broker.go:135
[BeforeEach] [service-catalog] Broker
  /home/jpeeler/src/go/src/github.com/kubernetes-incubator/service-catalog/test/e2e/framework/framework.go:53
STEP: Creating a kubernetes client
Apr 28 11:09:24.153: INFO: >>> config: /var/run/kubernetes/admin.kubeconfig

STEP: Creating a service catalog client
Apr 28 11:09:24.161: INFO: >>> config: /tmp/sc-kubeconfig

STEP: Building a namespace api object
Apr 28 11:09:24.162: INFO: namespace: &Namespace{ObjectMeta:k8s_io_apimachinery_pkg_apis_meta_v1.ObjectMeta{Name:,GenerateName:e2e-tests-create-broker-,Namespace:,SelfLink:,UID:,ResourceVersion:,Generation:0,CreationTimestamp:0001-01-01 00:00:00 +0000 UTC,DeletionTimestamp:<nil>,DeletionGracePeriodSeconds:nil,Labels:map[string]string{},Annotations:map[string]string{},OwnerReferences:[],Finalizers:[],ClusterName:,},Spec:NamespaceSpec{Finalizers:[],},Status:NamespaceStatus{Phase:,},}
[BeforeEach] [service-catalog] Broker
  /home/jpeeler/src/go/src/github.com/kubernetes-incubator/service-catalog/test/e2e/broker.go:108
STEP: Creating a user broker pod
STEP: Waiting for pod to be running
STEP: Creating a user broker service
[It] should become ready
  /home/jpeeler/src/go/src/github.com/kubernetes-incubator/service-catalog/test/e2e/broker.go:135
STEP: Creating a Broker
STEP: Waiting for Broker to be ready
[AfterEach] [service-catalog] Broker
  /home/jpeeler/src/go/src/github.com/kubernetes-incubator/service-catalog/test/e2e/framework/framework.go:54
[AfterEach] [service-catalog] Broker
  /home/jpeeler/src/go/src/github.com/kubernetes-incubator/service-catalog/test/e2e/broker.go:117
STEP: Deleting the user broker pod
STEP: Deleting the user broker service

• Failure [35.259 seconds]
[service-catalog] Broker
/home/jpeeler/src/go/src/github.com/kubernetes-incubator/service-catalog/test/e2e/framework/framework.go:89
  should become ready [It]
  /home/jpeeler/src/go/src/github.com/kubernetes-incubator/service-catalog/test/e2e/broker.go:135

  Expected error:
      <*errors.errorString | 0xc420355b20>: {
          s: "timed out waiting for the condition",
      }
      timed out waiting for the condition
  not to have occurred

  /home/jpeeler/src/go/src/github.com/kubernetes-incubator/service-catalog/test/e2e/broker.go:132
```

**Comment by jpeeler** [comment]
Ran 1 of 1 Specs in 5.750 seconds

Turns out there was some docker issue at play in the above error. Looks good!

**Comment by pmorie** [comment]
That is awesome

File: test/e2e/broker.go (hunk 2 above)
Comment (MHBauer) [inline test/e2e/broker.go]: ginkgo!

**Pull request merged**